### ISSUE TITLE
Client Lambda local invocation issue

### ISSUE THREAD

**Issue by itslior** [aws/aws-sdk-js-v3#6022]
### Describe the bug

TL;DR issue - 
1. During local flow of lambda A, when attempting to invoke lambda B locally results in an ARN resource error
2. When not using a `default` profile, it fails to identify the profile correctly.

How we encountered the bug?
When migrating our sdk in the serverless environment from `aws-sdk` V2 to V3, I followed the recommended migration documents, and when I reached the `client-lambda` related code, I replaced everything from V2 and updated it to V3 resulting in the code in the reproduction segment.

Additionally, when not using the`default` profile or any other indicator (such as `process.env.AWS_PROFILE` value) it fails, as the `init.profile` within the `lambda.send` flow (I believe it's used in conjuction with `@smithy`) is not defined well

### Expected Behavior

1. The lambda is invoked locally successfully with the default credentials stored.
2. When using a non default value for the profile - it should work and pick up the credentials as expected

### Current Behavior

1. 
× Unhandled exception in handler 'mockFunctionA'.
× ResourceNotFoundException: Function not found: arn:aws:lambda:<redacted>:<redacted>:function:test-service-dev-push-metric-values

2. 
× Unhandled exception in handler 'mockFunctionA'.
× CredentialsProviderError: Could not load credentials from any providers

### Reproduction Steps

Create a basic serverless repo with two two functions, have one function be triggered by an HTTP event and during it's flow invoke function B.
Worth to note that I have the sufficient plugins installed and worked flawlessly in SDK V2

```
import { fromIni } from '@aws-sdk/credential-providers';
import { InvokeCommand, LambdaClient } from '@aws-sdk/client-lambda';
...
const lambda = new LambdaClient([{
  endpoint: isLocal ? 'http://localhost:4000' : 'https://lambda.us-east-1.amazonaws.com';
  credentials: fromIni({ profile: 'default' }),
}]);
  const payload = {
  metricValueArray: ['testValue'],
};
await lambda.send(new InvokeCommand({
    FunctionName: 'test-service-dev-push-metric-values',
    InvocationType: 'Event',
    LogType: 'Tail',
    Payload: JSON.stringify(payload),
}));
```

The versions I'm using:
```
"serverless-offline": "^12.0.4",
"@aws-sdk/client-lambda": "^3.521.0",
"@aws-sdk/credential-providers": "^3.521.0",
```

2. to reproduce issue 2 that's listed previously, just change the `default` profile to anything else basically (this option is practically not used anyway)

### Possible Solution

I managed to get a work around which also leads me to suggest another adjustment (I don't mind opening the PR assuming I didn't get anything wrong)

Here it is:
```
const endpoint: Provider<EndpointV2> = () => new Promise((resolve) => {
  const ep: EndpointV2 = {
    url: new URL(isLocal ? 'http://localhost:4000' : 'https://lambda.us-east-1.amazonaws.com'),
  };
  resolve(ep as any);
});
const lambda = new LambdaClient();
lambda.config.credentials = fromIni({ profile: 'default' });
lambda.config.endpoint = endpoint as any;
const payload = {
  metricValueArray: ['testValue'],
};
await lambda.send(new InvokeCommand({
    FunctionName: 'test-service-dev-push-metric-values',
    InvocationType: 'Event',
    LogType: 'Tail',
    Payload: JSON.stringify(payload),
  }));
```

The versions used are the same as in the snippets provided above.
- Worth to note that if you have this line:
`lambda.config.credentials = isLocal ? fromIni({ profile: 'default' }); : undefined;`
it wont work either and instead you need to do 
```
if (isLocal) {
    lambda.config.credentials = fromIni({ profile: 'default' });
}
```
but this works
```
const lambda = new LambdaClient([{
  endpoint: isLocal ? 'http://localhost:4000' : 'https://lambda.us-east-1.amazonaws.com';
  credentials: isLocal ? fromIni({ profile: 'default' }) : undefined,
}]);
```

Quick explanation:
I tried using the `Endpoint` type (which is deprecated) as it should be, as `lambda.config.endpoint` is of `Endpoint` type, however, trying to configure it I ran into issues (which is a separate manner) and instead, I figured that if the base config (which I assumed you fall back to) is able to accept `EndpointV2` I basically shifted it to be of that type and forced it as any (not ideal), but testing that theory proved to work and the lambda was successfully invoked, which leads me to also suggest that you add an `| Provider<EndpointV2>` to the type of the `lambda.config.endpoint` type, and also, leads me to suspect that the fallback to the base config when the `LambdaClient` is defined, isn't happening properly, attached below:

```
/**
 * @internal
 *
 * This supercedes the similarly named EndpointsResolvedConfig (no parametric types)
 * from resolveEndpointsConfig.ts in @smithy/config-resolver.
 */
export interface EndpointResolvedConfig<T extends EndpointParameters = EndpointParameters> {
    /**
     * Custom endpoint provided by the user.
     * This is normalized to a single interface from the various acceptable types.
     * This field will be undefined if a custom endpoint is not provided.
     */
    endpoint?: Provider<Endpoint> | Provider<EndpointV2>; // <---- This is the modification that I'm suggesting ontop of solving falling back to default's initialized issue I described
    endpointProvider: (params: T, context?: {
        logger?: Logger;
    }) => EndpointV2;
    /**
     * Whether TLS is enabled for requests.
     * @deprecated
     */
    tls: boolean;
    /**
     * Whether the endpoint is specified by caller.
     * @internal
     * @deprecated
     */
    isCustomEndpoint?: boolean;
    /**
     * Resolved value for input {@link EndpointsInputConfig.useDualstackEndpoint}
     */
    useDualstackEndpoint: Provider<boolean>;
    /**
     * Resolved value for input {@link EndpointsInputConfig.useFipsEndpoint}
     */
    useFipsEndpoint: Provider<boolean>;
    /**
     * Unique service identifier.
     * @internal
     */
    serviceId?: string;
}

```

Regarding issue number 2 listed above - just use the `process.env.AWS_PROFILE` option

### SDK version used

3.521.0

### Environment details (OS name and version, etc.)

Mac Sonoma 14.3.1

**Comment by itslior** [comment]
Hey @RanVaknin 
Just making sure - you're saying that if I don't use the default profile, then I need to, in the `fromIni` function also state the region, even if it is the default `us-east-1` region?
Asking cause locally we work with a profile that isn't the default, but we do use the default region.

**Comment by theoffricial** [comment]
Hi @RanVaknin thanks for responding us, we did so many tests and also reverse engineering. I hope the configuration you provided will work for us, but I have to say that it isn't intuitive at all and the aws-sdk V3 doesn't come with JSDocs or documentation that points we should use it, maybe adding clarifications on your docs can help many people to adopt it.
Thanks

**Comment by kuhe** [comment]
`clientConfig` is only used if an additional client is needed to fetch credentials, such as STS, SSO, or SSO-OIDC. The `clientConfig` object is passed to these clients. 

If you have static credentials, `fromIni` will not use the `clientConfig` parameter, because no additional client is used to resolve the credentials. 

The region of a client is defined separately, even if your selected profile has a region associated with it. 
The order of precedence for resolving the region is:

1. in code for client initialization.

```ts
new Client({
  region: "us-east-1",
});
```

2. fromEnv
```
AWS_REGION=us-east-1 node your_program.js
```

3. from the config file profile defined by AWS_PROFILE or `default`. 

Specifically for number 3, the profile selected by `fromIni` is not used to resolve the region, only the credentials. If you want to use a profile's region, you must set `AWS_PROFILE` on the env.

**Comment by itslior** [comment]
I think there might be a misunderstanding, so just to make sure @kuhe -
When using aws-sdk-2 and the parallel function of what `fromIni` was -> `SharedIniFileCredentials`, when you passed a profile that isn't the default profile to the `profile` param, it chose the profile from the local credentials file that was created prior to running the flow (from clisso for example).
When using aws-sdk-3, `fromIni` ignores the profile param that you pass onto it and only pulls from `default` or `AWS_PROFILE`, basically, the issue is that `profile` param that `fromIni` expects to get is ignored.

That's the issue with one of the issues I presented above

**Comment by RanVaknin** [comment]
Hi @itslior ,

Thanks for your patience. I confirmed what @kuhe said in his [comment](https://github.com/aws/aws-sdk-js-v3/issues/6022#issuecomment-2098745076). The fromINI provider does not read the region from the config file. It is only loading credentials. In v3 you'll have to provide the region in one of the ways provided in the comment.

 George was also right regarding the workaround I provided. providing `clientConfig` is only applicable for other credential providers that call the internal STS client under the hood. 

I'll review this with the team.
Thanks,
Ran~

**Comment by theoffricial** [comment]
@kuhe @RanVaknin The issue @itslior describes is that when we tried to configure the Lambda constructor using its constructor settings, it didn't work as we expected and didn't configure it properly. After some trial and error, we found the setup to work, as @itslior described, but it wasn't straightforward and required us to reverse-engineer the AWS-SDK packages and `@smithy` packages.

@itslior is the answer provided by @kuhe to clarify the gap we detected in the Lambda invocation setup?
